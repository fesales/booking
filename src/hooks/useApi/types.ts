export enum FetchStatus {
  ERROR = 'ERROR',
  PENDING = 'PENDING',
  SUCCESS = 'SUCCESS',
}

export interface FetchError {
  message: string;
}

interface PendingState {
  response: null;
  status: FetchStatus.PENDING;
}

export interface SuccessState<ApiResponse> {
  response: ApiResponse | null;
  status: FetchStatus.SUCCESS;
}

interface ErrorState {
  response: FetchError;
  status: FetchStatus.ERROR;
}

export type State<ApiResponse> = PendingState | SuccessState<ApiResponse> | ErrorState;

export enum Actions {
  REQUEST_LOADING = 'REQUEST_LOADING',
  REQUEST_SUCCESS = 'REQUEST_SUCCESS',
  REQUEST_ERROR = 'REQUEST_ERROR',
}

export interface RequestLoadingAction {
  type: Actions.REQUEST_LOADING;
}

export interface RequestSuccessAction<ApiResponse> {
  payload: ApiResponse;
  type: Actions.REQUEST_SUCCESS;
}

export interface RequestErrorAction {
  payload: FetchError;
  type: Actions.REQUEST_ERROR;
}

export type Action<ApiResponse> =
  | RequestErrorAction
  | RequestSuccessAction<ApiResponse>
  | RequestLoadingAction;
