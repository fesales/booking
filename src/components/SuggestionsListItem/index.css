.suggestions-list-item {
  border-bottom: 1px solid #b4b4b4;
  display: flex;
  cursor: pointer;
  font-size: 12px;
  padding: 8px 16px;
}

.suggestions-list-item:hover {
  background-color: #e7f4fe;
}

.suggestions-list-item-name {
  display: block;
  font-size: 14px;
}

.suggestion-list-item-place {
  border-radius: 4px;
  color: rgb(255, 255, 255);
  font-weight: 600;
  height: fit-content;
  line-height: 21px;
  margin-right: 25px;
  min-width: 60px;
  padding: 1px 4px;
  text-align: center;
}

.suggestion-list-item-place--A {
  background-color: rgb(150, 20, 18);
}

.suggestion-list-item-place--C {
  background-color: rgb(10, 99, 176);
}

.suggestion-list-item-place--T {
  background-color: rgb(91, 91, 91);
}

.suggestions-list-item-place-wrapper {
  display: inline-flex;
}
