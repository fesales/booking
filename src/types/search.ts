export enum PlaceType {
  AIRPORT = 'A',
  CITY = 'C',
  STATION = 'T',
}

export interface PickUpLocation {
  city: string;
  country: string;
  name: string;
  region: string;
  placeType: PlaceType;
}

export interface SearchApiResponse {
  results: {
    isGooglePowered: boolean;
    docs: PickUpLocation[];
    numFound: number;
  };
}
