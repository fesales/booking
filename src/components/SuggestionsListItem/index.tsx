import React, { useMemo } from 'react';
import classnames from 'classnames';
import './index.css';
import { PickUpLocation, PlaceType } from '../../types/search';
import PLACE_TYPES_NAMES from '../../constants/place-type';

const SuggestionsListItem = ({ city, country, name, placeType }: PickUpLocation) => {
  const locationInfo = useMemo(() => {
    return city ? `${city || ''}${country ? `, ${country}` : ''}` : country;
  }, [city, country]);

  const placeTypeClass = classnames({
    'suggestion-list-item-place': true,
    'suggestion-list-item-place--A': placeType === PlaceType.AIRPORT,
    'suggestion-list-item-place--C': placeType === PlaceType.CITY,
    'suggestion-list-item-place--T': placeType === PlaceType.STATION,
  });

  return (
    <li className="suggestions-list-item">
      <div className="suggestions-list-item-place-wrapper">
        <span className={placeTypeClass}>{PLACE_TYPES_NAMES[placeType]}</span>
      </div>
      <div>
        <span className="suggestions-list-item-name">{name}</span>
        <span className="suggestions-list-item-info">{locationInfo}</span>
      </div>
    </li>
  );
};

export default SuggestionsListItem;
