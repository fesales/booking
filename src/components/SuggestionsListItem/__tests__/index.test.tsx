import React from 'react';
import { shallow } from 'enzyme';
import SuggestionsListItem from '..';
import { PlaceType, PickUpLocation } from '../../../types/search';

describe('SuggestionsListItem', () => {
  const defaultProps: PickUpLocation = {
    city: 'Manchester',
    country: 'United Kingdom',
    name: 'pick up location',
    placeType: PlaceType.AIRPORT,
    region: '',
  };

  it('should render component', () => {
    const wrapper = shallow(<SuggestionsListItem {...defaultProps} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.suggestions-list-item-name').text()).toEqual(defaultProps.name);
    expect(wrapper.find('.suggestion-list-item-place--A').text()).toEqual('Airport');
    expect(wrapper.find('.suggestions-list-item-info').text()).toEqual(
      'Manchester, United Kingdom'
    );
  });

  describe('location class name', () => {
    it('should add airport class', () => {
      const wrapper = shallow(<SuggestionsListItem {...defaultProps} />);

      expect(wrapper.find('.suggestion-list-item-place--A').text()).toEqual('Airport');
    });

    it('should add city class', () => {
      const props = {
        ...defaultProps,
        placeType: PlaceType.CITY,
      };

      const wrapper = shallow(<SuggestionsListItem {...props} />);

      expect(wrapper.find('.suggestion-list-item-place--C').text()).toEqual('City');
    });

    it('should add station class', () => {
      const props = {
        ...defaultProps,
        placeType: PlaceType.STATION,
      };

      const wrapper = shallow(<SuggestionsListItem {...props} />);

      expect(wrapper.find('.suggestion-list-item-place--T').text()).toEqual('Station');
    });
  });
});
