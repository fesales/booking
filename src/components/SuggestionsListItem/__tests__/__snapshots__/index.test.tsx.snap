// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuggestionsListItem should render component 1`] = `
<li
  className="suggestions-list-item"
>
  <div
    className="suggestions-list-item-place-wrapper"
  >
    <span
      className="suggestion-list-item-place suggestion-list-item-place--A"
    >
      Airport
    </span>
  </div>
  <div>
    <span
      className="suggestions-list-item-name"
    >
      pick up location
    </span>
    <span
      className="suggestions-list-item-info"
    >
      Manchester, United Kingdom
    </span>
  </div>
</li>
`;
